Fill in kubelet raw metrics API

timstclair opened the pull request:
This is an initial specification of the raw metrics kubelet API. The
initial customer of the API will be Heapster, so the fields have been
restricted to those that are necessary to move Heapster off the old
/stats endpoint.

Issue: https://github.com/kubernetes/kubernetes/issues/12483
Proposal (modified to match this PR): https://github.com/timstclair/kubernetes/blob/metrics-beta2/docs/proposals/compute-resource-metrics-api.md

@vishh @bgrant0607 @kubernetes/goog-node @mwielgus @jszczepkowski @piosz @jimmidyson

tallclair committed:
Fill in kubelet raw metrics API

This is an initial specification of the raw metrics kubelet API. The
initial customer of the API will be Heapster, so the fields have been
restricted to those that are necessary to move Heapster off the old
/stats endpoint.

jimmidyson commented on pkg/apis/metrics/v1alpha1/types.go:
```diff
@@ -16,14 +16,208 @@ limitations under the License.
+	TotalCores *resource.Quantity `json:"totalCores,omitempty"`
+	// CPU load that the container is experiencing, represented as a smoothed average of number of
+	// runnable threads.  Load is averaged over the sampling window.
+	LoadAverage *resource.Quantity `json:"loadAverage,omitempty"`
```
Can we actually provide load average on a per-container basic? How is this calculated?

timstclair replied:
`CPUMetrics` are included in `ContainerSample`, which is included in `RawContainer`, so they are already per-container. Is that what you meant?

As for how it's calculated, I think the comment may actually be innacurate. I think it's the number of **running tasks**, not the number of "runnable threads". It's read from the kernel CPU stats: https://github.com/google/cadvisor/blob/master/info/v1/container.go#L241

jimmidyson replied:
Right so my point is can you calculate load on a per-container basis at all? I'm not convinced you can but I'm open to learning how.

timstclair replied:
Ah, interesting, I think you are correct. I need to stop assuming that because something is in the cAdvisor API we have access to it :)

I think that TotalCores is sufficient for now, so I removed this. I don't think it was being computed quite right anyway...

jimmidyson commented on docs/proposals/compute-resource-metrics-api.md:
```diff
@@ -115,13 +116,11 @@ only be available through the kubelet. The types of metrics are detailed
-  `time.Time`; since samples are retrieved at discrete intervals, the first
-  sample after the start time is the actual beginning.
-- `end` - end time to return metrics to; type json encoded `time.Time`
-- `step` - the time step between each stats sample; type int (seconds), default
```
What's the reason behind dropping `step` parameter? Found this very useful when working with Prometheus to get smoother metrics responses when the time window is larger.

timstclair replied:
Basically it looks like we don't need it for our initial usecase (heapster). This may be more useful for the DerivedMetrics, and I'm not against adding it back. Do you anticipate it being used with RawMetrics?

mwielgus replied:
Without step or even samples spreading maxSamples > 1 is not particularly useful. Imagine 1 min window and maxSamples set to 3. Having samples at 20s, 40s and 60s (step=20) is much more useful than at 58, 58, 60s. So either drop it entirely (Heapster will most likely set it to 1 anyway) or add step.

timstclair replied:
Ok, I can add this back. What is the desired behavior of step? E.g. with a 10s sampling window, and a 30s step, should it drop samples between the 30s samples, or take the 3 sets and average them? If we want to average, what should we do when data is partial (i.e. missing from some samples)?

mwielgus replied:
If we only have a poll api (for now) then a step bigger than a sampling window should mean that only 1 sample is returned.

jimmidyson replied:
Sorry (not very) to keep using Prometheus as a guide, but what Prometheus does is provide two queries: range & instantaneous. Do we see a need to add this? Using instantaneous queries is very useful for getting most recent metrics in a time window (e.g. now).

timstclair replied:
I think the parameters I've provided here are a superset of the prometheus instantaneous & range queries. If I understood correctly, an instantaneous query is equivalent to setting `maxSamples=1`, and a range query is equivalent to setting `sinceSeconds`? I'm still not sure how step should work, though I'm inclined to just have it drop samples that fall between steps.

jimmidyson replied:
Agreed.

Yes step should drop samples between steps IMO.

timstclair replied:
Another question: Should the most recent metric always be included, and then "step" backwards from there? Or should the step be from some fixed time in the past (e.g. epoch)?

For example,
Step from most recent sample:
```
Input: start:=100 end=201 step=30
==> Output: samples[200, 170, 140, 110]
Input start=100 end=211 step=30
==> Output: samples[210, 180, 150, 120]
```
vs. Step from fixed time (0 in this example):
```
Input: start:=100 end=201 step=30
==> Output: samples[180, 150, 120]
Input start=100 end=211 step=30
==> Output: samples[210, 180, 150, 120]
```

mwielgus replied:
I think that we should start from the most recent sample/end of the query window. Starting from the fixed time 0 would be hard to understand, especially if step is not a divisor of 60. 

The other possibility is return the newest sample from  the step-sized window, starting from end. In your example it would be:
start=100 stop=201 step=30  => [200, 170, 140, 110]
 or if there were some "holes" [197, 169, 135, 105]
This will prevent going completely out of sync in case some missing metrics/hiccups.

timstclair replied:
Ok, from some offline discussions, there is a counter argument for leaving out `step` and also dropping `start` and `end` (`since / until`): The node is only configured to keep 1-2 minutes of historical data to keep memory usage down, which limits the usefulness of those stats. Furthermore, clocks may be out of sync across nodes, so some additional client complexity is necessary to properly use start & end anyway.

Along the same lines, we should not be logging data frequently enough (except in some special circumstances) that `step` should matter much (e.g. if you want 20s intervals, just request the default 10s intervals, and drop every other sample).

I'm inclined to start with this in order to unblock this PR and simplify the implementation, since it seems to meet our current MVP use case. Remember that this API can grow after this PR - this is just an initial version.

/discuss :)

jimmidyson commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+type CustomMetricSample struct {
+	Sample `json:",inline"`
+	// Label associated with a metric
+	Label *string `json:"label,omitempty"`
```
Should this be a `map[string]string` as well?

timstclair replied:
No. It's actually a little unclear to me what exactly this label is, or how it's used. I believe it's just the first key:value pair parsed from the prometheus metrics, but that seems a little odd. I'll update this once I track it down.

dchen1107 replied:
@rjnagal on custom metrics.

jimmidyson replied:
I'd actually like to drop custom metrics as I see it as adding resource use (duplicate memory for holding application metrics in both application & cadvisor) for little value that I can see.

jimmidyson commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	// Historical metric samples gathered from the container.
+	Samples []ContainerSample `json:"samples" patchStrategy:"merge" patchMergeKey:"sampleTime"`
+	// Application-specific metrics.
+	CustomMetrics []CustomMetric `json:"customMetrics,omitempty" patchStrategy:"merge" patchMergeKey:"name"`
```
Do we want to actually include custom metrics in alpha1? Seems to me that there is quite a bit of stuff to sort out before we can actually collect custom metrics, unless we use Prometheus which does this already. Should we drop custom metrics for this proposal & extend it later as we've figured it out?

timstclair replied:
I was told this was a requirement for moving Heapster off of the `/stats` endpoint. Maybe @mwielgus can provide more detail about how it's used.

mwielgus replied:
Custom metrics support has to be placed in the api one way or another as this is one of the key deliverables of the WAW K8S team. They are quite well defined in the cAdvisor structs and I would suggest following that pattern. As this is the alpha api and Heapster is the only client we are free to change it when/if any better design appear.

mwielgus replied:
@jszczepkowski

jimmidyson replied:
Prometheus provides this today without the duplicate resource usage described above.

timstclair replied:
What do you mean by "duplicate resource usage"?

jimmidyson replied:
No I'm not concerned with Heapster (or whatever is the source for HPA data) storing them.

My concern is that while running applications will be keeping metrics in memory in order to be scraped, by Prometheus or by cadvisor, cadvisor will also be keeping those metrics in memory to be scraped by Heapster (or other system). Obviously we need to ensure that memory footprint for anything running on nodes is kept to a minimum to get max resources available for applications. Anything we can do to achieve that we should do IMO.

mwielgus replied:
We are talking about couple custom metrics per pod/container. Tha'ts couple ints and strings and  the number of custom metrics can be limited to some reasonable value. I don't think that killing features for the sake of micro-optimizations is the right thing to do. There are probably other places where the memory can be saved in much greater quantity.

jimmidyson commented on pkg/apis/metrics/v1alpha1/types.go:
```diff
@@ -16,14 +16,208 @@ limitations under the License.
+	// Instantaneous value. May increase or decrease.
+	CustomMetricGauge CustomMetricType = "Gauge"
+	// A counter-like value that increases monotonically.
+	CustomMetricCumulative CustomMetricType = "Cumulative"
```
Would we be better just reusing the Prometheus metric types as these are pretty well known now. `Counter` instead of `Cumulative` & `Rate` instead of `Delta`?

mwielgus replied:
This is an api on top of cAdvisor and cAdvisor uses Gauge, Cumulative and Delta. Sure, the names could be better but I think we should be rather consistent with cAdvisor than with Prometheus.

timstclair replied:
I don't have an opinion either way on this issue. If as Jimmi says the prometheus names are becoming standard, it may make sense to rename the cAdvisor types to match (in a V3 API, if we ever do that).

jimmidyson replied:
Prometheus metric types & exposition format is definitely gaining popularity.

`Cumulative` & `Delta` are not descriptive nouns of what they represent so semantically I'd prefer `Counter` & `Rate`, even if they weren't used in Prometheus.

jimmidyson commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+type CPUMetrics struct {
+	// Total CPU usage (sum of all cores) averaged over the sample window.
+	// The "core" unit can be interpreted as CPU core-seconds per second.
+	TotalCores *resource.Quantity `json:"totalCores,omitempty"`
```
Without knowing number of cores available this isn't that useful, is it?

timstclair replied:
I'm trying to avoid duplicating data that's provided by the kubernetes API. In this case, number of cores is provided by Capacity (will be [Allocatable](https://github.com/kubernetes/kubernetes/pull/17201)).

jimmidyson replied:
OK so in this case the percentage would be calculated on the API server? Can allocated cores change over time? Do we need to track usage with allocation in same sample?

timstclair replied:
Hmm, interesting point about correlating historical `Allocatable` resources with the usage - I think this is a good reason *not* to report a percentage though. Allocatable can change, but I don't expect it to fluctuate too drastically in the normal use cases.

I'm not sure what you mean by "the percentage would be calculated on the API server"? I don't think so... where is the percentage needed?

jimmidyson replied:
Isn't percentage needed by HPA?

timstclair replied:
I believe HPA uses the percantage of utilization from the requested CPU-cores (as opposed to capacity). Request (and limit) are reported by Heapster alongside utilization, so I don't think we need to do anything here.

jimmidyson commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+// NetworkMetrics contains data about network resources.
+type NetworkMetrics struct {
+	// Cumulative count of bytes received.
+	RxBytes *resource.Quantity `json:"rxBytes,omitempty"`
```
Can you explain the preference for `resource.Quantity` everywhere to raw numbers? De-serializing `Quantity` introduces logic that every client would have to implement which is a bit annoying. I'd prefer to use `int64` or `float64` everywhere.

The usage also seems inconsistent - can you explain why some properties are `int64` & some are `resource.Quantity`?

jimmidyson replied:
I assume you're using `resource.Quantity` to be able to represent floats? We could do the same just using `float64`, albeit with the lower max precision, but even the max `float64` is going to be enough for every use case IMO. This would simplify client libraries & make the API more consistent.

timstclair replied:
Floats are discouraged in the API: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#primitive-types
It might not actually matter in this context since all these values are read-only, but it is less consistent. For non-fractional values (e.g. bytes) I think the decision is a little more arbitrary. My concern is that a high-traffic long-lived node could exceed the max int value (~9 petabytes).

It's a good point about the additional development complexity of integrating resource.Quantity, it might be a good thing to move to a separate repo (of course that will only help go implementations, although if it's in a separate repo we could provide alternate implementations as well).

jimmidyson replied:
Interesting note about converting to strings for values > 53 bits. So that's the reason behind Quantity - never knew that. I do like the idea of a separate repo with multiple language implementations - this has uses outside of k8s.

mwielgus commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+	UntilTime *unversioned.Time `json:"untilTime,omitempty"`
+	// Specifies the maximum number of elements in any list of samples.
+	// When the total number of samples exceeds the maximum the most recent samples are returned.
+	MaxSamples int `json:"maxSamples,omitempty"`
```
For maxSample > 1 there should be a step so that 1 can specify that 1 want 3 samples from the last minute, spaced by 20 sec. Returning 3 most recent samples for 1 min window is not particularly useful.

timstclair replied:
Ack. Responded above.

mwielgus commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,205 @@ limitations under the License.
+}
+
+// RawMetricsOptions are the query options for raw metrics endpoints.
+type RawMetricsOptions struct {
```
If you are planning to provide a Watch-like api then there should be a way to express when to send the next set of metrics. I would like to aviod the situation when all Kubelets push their new data to Heapster at the same time.

timstclair replied:
Agreed. I'm thinking we should stick with the polling API for 1.2, based on a previous discussion on this point.

jimmidyson replied:
This is why pulling metrics is a better approach IMO - overloading the server with pushes at the same time is a real danger, requiring back pressure, extra monitoring of each node to ensure pushes have happened successfully & that back pressure is working appropriately.

tallclair committed:
Add FilesystemMetrics back

timstclair commented:
Added back FilesystemMetrics, which are also included in Heapster data.

tallclair committed:
Remove CPUMetrics.LoadAverage

jimmidyson commented:
This is very close & we all want to get this moving so how about we remove custom metrics for `v1alpha1` & get raw metrics working while we discuss v1alpha2, v1beta1, even v1 :)

mwielgus commented:
As I mentioned in the code review, custom metrics are really important to us. The new Heapster won't switch to Metrics API if there is no custom metrics support. We want to have SOME custom metrics from day 1, even if we have to refactor our code couple days/weeks later, when you come up with the new shiny design.

mwielgus commented:
cc: @fgrzadkowski

jszczepkowski commented:
> This is very close & we all want to get this moving so how about we remove custom metrics for v1alpha1 & get raw metrics working while we discuss v1alpha2, v1beta1, even v1 :)

Custom metrics in Kubernetes (#17431) and autoscaling pods on base of them is one of key 1.2 features. Removing them from 1.2 is not an option.

jimmidyson commented:
OK fair enough @mwielgus @jszczepkowski. I guess my problem is really with the implementation rather than the API anyway so :+1:

jimmidyson commented on pkg/apis/metrics/types.go:
```diff
@@ -16,14 +16,216 @@ limitations under the License.
+	// Metrics pertaining to memory (RAM) resources.
+	Memory *MemoryMetrics `json:"memory,omitempty"`
+	// Metrics pertaining to network resources.
+	Network *NetworkMetrics `json:"network,omitempty"`
```
Do we care about multiple interfaces? Guess just total in/out is enough.

timstclair replied:
I don't think per-interface stats are exposed by heapster (correct me if I'm wrong). My only concern about only including the aggregate stats here is that we probably don't want to include stats from the loopback and other virtual interfaces.

I would appreciated more input on this issue. @vishh @dchen1107

timstclair commented:
> OK fair enough @mwielgus @jszczepkowski. I guess my problem is really with the implementation rather than the API anyway so :+1:

Agreed. Since the API is served by the Kubelet and is not tied to cAdvisor, we could gather the custom metrics from elsewhere if we decide to disable custom metrics in cAdvisor.

timstclair commented:
@jszczepkowski 

> Custom metrics in Kubernetes (#17431) and autoscaling pods on base of them is one of key 1.2 features. Removing them from 1.2 is not an option.

From an offline discussion, it sounds like there are still some open questions around custom metrics (e.g. are they per-container or per-pod?). I'm inclined to follow @jimmidyson 's advice and drop custom metrics from this initial version. We can add them as soon as a more concrete proposal is in place. This will help unblock this API, so we can push forward with the well defined pieces.

tallclair committed:
Remove CustomMetrics from initial API.

More discussion is needed around CustomMetrics, including:

- Are custom metrics per-pod, per-container, or both?
- How are custom metrics structured? The kubelet API should match the
  heapster API

To unblock the rest of this change and give the discussion around custom
metrics the attention it deserves I'm removing custom metrics from this
initial version, to be added back once the above issues are resolved.

timstclair commented:
More discussion is needed around CustomMetrics, including:

- Are custom metrics per-pod, per-container, or both?
- How are custom metrics structured? The kubelet API should match the
  heapster API

To unblock the rest of this change and give the discussion around custom
metrics the attention it deserves I'm removing custom metrics from this
initial version, to be added back once the above issues are resolved.

CustomMetrics discussion here: https://github.com/kubernetes/kubernetes/pull/17431

tallclair committed:
Remove {since,until}Time from RawMetricsOptions

In the interest of simplifying this PR and unblocking the basic
implementation, I'm removing extra parameters which are not strictly
necessary for v1alpha1. This is not to say these parameters will never
be added, but rather just to unblock this initial implementation.

tallclair committed:
Fix v1alpha1 code gen

erictune commented:
Why expose an interface on kubelet that has all these metrics, when the only one needed for HPA is cpu?

timstclair commented:
@erictune - To clarify, HPA was the motivation for doing this work, but there is an additional goal of supporting all of Heapster's needs so that we can deprecate the `/stats` endpoint, which currently exports WAY more data (ex: [`/stats` response](https://gist.github.com/timstclair/1993260c0e5b6ee09563))

vishh commented:
Heapster needs to be scalable for HPA GA. This API is an effort to help with heapster scalability.

mwielgus commented:
@erictune The goal is to use only the new api in Heapster. So other metrics should go there as well.

erictune commented:
Why is it a requirement to continue to export all the same data to heapster that we were before?

jimmidyson commented:
HPA is just one consumer of metrics. For monitoring purposes all the other metrics are required, e.g. find out which pods are using up all network bandwidth, etc.

Another option would be that the kubelet just serves the metrics required for cluster component, e.g HPA, & another process serves richer metrics? This seems to introduce unnecessary deployment complexity.

jimmidyson commented:
Alternatively we could use the already exposed Prometheus endpoint at `/metrics` which provides much more granular metrics. But I prefer having the kube semantics for querying kubelet for metrics generally.

mwielgus commented:
@erictune Sure, we can drop everything except cpu-usage from Heapster. And maybe even drop HPA along cpu-usage and then kill Heapster, Grafana and integration with GKE. But why?

A user who have been using Kuberentes for half a year would feel at least disappointed if suddenly we stopped passing the metrics for no reason. If you don't like one particular metric - sure we can drop it or replace with some other. But I strongly feel that the current Kubernetes functionality should be kept more or less as is.

jszczepkowski commented:
@timstclair 
Regarding custom metrics, I think we definitely should place them in kubelet metrics API ASAP.

> Are custom metrics per-pod, per-container, or both?

In cadvisor, they are reported per-container (as everything). However, in k8s we have them per-pod. In my opinion, kubelet should expose them to heapster per-pod (the sooner we convert them to be pre-pod, the better).

>  "Finalizing" custom metrics design (i.e. merge #17431)

Done.

>  Stress test custom metrics implementation, specifically performance impacts

I done some performance tests for custom metrics. I put the results (output of top for kubelet) here:

```
PID   USER      PR  NI   VIRT  RES SHR S  %CPU %MEM  TIME+    COMMAND

Empty cluster (few minutes after creation):
 3306 root      20   0  131m  38m  25m S   0.0  0.5   0:02.49 kubelet
 3358 root      20   0  139m  38m  25m S   6.6  0.5   0:02.15 kubelet
 3305 root      20   0  140m  39m  25m S   0.0  0.5   0:02.70 kubelet

30 pods per node, no custom metrics:
 3306 root      20   0  196m  78m  25m S  13.1  1.1   0:50.46 kubelet
 3358 root      20   0  196m  81m  26m S  13.0  1.1   0:43.38 kubelet
 3305 root      20   0  207m  84m  25m S   6.5  1.1   0:47.97 kubelet

30 pods per node, 5 custom metrics per pod:
 3728 root      20   0  234m 109m  25m S   6.5  1.5   2:53.90 kubelet
 3738 root      20   0  227m 110m  25m S  19.6  1.5   3:07.14 kubelet
 3729 root      20   0  214m 107m  26m S   6.5  1.4   2:56.48 kubelet

30 pods per node, 10 custom metrics per pod:
 3753 root      20   0  237m 104m  26m S  19.5  1.4   2:40.88 kubelet
 3729 root      20   0  260m 116m  25m S   6.5  1.5   2:43.24 kubelet
 3745 root      20   0  227m 110m  25m S  13.1  1.5   2:39.92 kubelet

``` 
(Statistics were gathered after about ~30 minutes, when top results were stable).

**Conclusions**: when custom metrics are gathered, memory consumption of kubelet is higher for about 25%. In my opinion, this is perfectly fine for 1.2 (we have limit for 5 custom metrics per pod):
* custom metrics are alpha feature, observed memory increase is acceptable,
* the increase should be much smaller due to #18044 (maybe even by order of magnitude),
* we may still do some tuning.

dchen1107 commented:
Had offline discussion with HPA folks, @timstclair earlier. 

1) HPA requires custom metrics, and it is in alpha release but a blocker for 1.2 release. Agreed to unblock them to make progress. 
2) On another side, when @vishh and I initially discussed with HPA folks on custom metrics support, we had big concern on who should be charged with the extra resource consumption.  The different user has different set of metrics, and consume different size of resource. We proposed several architectures. One of them is introducing a sidecar container for pod to collect such metrics and push to heapster.  But that requires much more work. Unfortunately, I couldn't find this from our meeting notes and current design doc as an alternative. At the end, to unblock this alpha feature, we agreed with current approach as an experimental, but want to limit at most 5 metrics per pod. @vishh, please correct me if I missed anything here? @bgrant0607 
3) We are changing the interval of cadvisor on collecting cgroup's metrics, but that interval might not apply to the custom metrics too. @jimmidyson 

Meanwhile, Kubelet Raw Metrics API we introduced here is not targeted for HPA initially. It is the steps for improving kubelet's reliability and heapster's scalability in general. It also provides a clean way to query system and daemons' resource usage on the node. The feature is beta feature here, and a blocker for 1.2 release. @erictune

Here is my proposed decision: 
Let's remove custom metrics from this PR, furthermore, we should remove custom metrics completely from proposed endpoint. Instead let introduce a new endpoint for custom metrics, and guard this feature by a flag (by default: off). In this case, HPA folks can take over custom metrics support completely, we won't block each other. Is this ok? 
 
@mwielgus @piosz @@jszczepkowski @fgrzadkowski

vishh commented:
+1 to @dchen1107's comments. To elaborate a bit more, @dchen1107's proposal is to add an alpha opt-in API in the kubelet, which will expose a combination of the custom metrics and the system metrics. 
Meanwhile @timstclair can work towards completing the beta API for system metrics through this PR.
Heapster will have to deal with these two separate endpoints. But, given that custom metrics don't have a standard schema, I'd expect heapster core to be resilient to lack of custom metrics.
The primary motivation behind this proposal (a new alpha API for custom metrics) is kubelet stability and resource overhead. There are several issues/feature requests currently filed for lowering kubelet's resource overhead. 
The hope is that by providing means to use custom metrics for HPA, users can try it out and help stabilize the custom metrics stack before graduating it to beta. 
@mwielgus @jszczepkowski let us know if there is a mis-understanding on the requirements front.

timstclair commented:
Just to be clear, I want separate endpoints for CustomMetrics within an API version, not just separate version endpoints. I'd like to see CustomMetrics endpoints mirror the RawMetrics endpoints, with `s/raw/custom/`. E.g.
```
/apis/metrics/v1alpha1/customNodes/localhost
/apis/metrics/v1alpha1/namespaces/{namespace}/customPods/{pod}
...
```

mwielgus commented:
Sorry but I don't agree with the idea of 2 separate endpoints. The main cost of custom metrics is in gathering them and storing in Kubelet, not in the API itself. Why can't we just have a field in the main api that will optionally have custom metrics if they are enabled on the kubelet? What are we trying to optimize? From the performance point of view it doesn't make sense to send 2 requests instead of one.

timstclair commented:
Closing this due to change in direction. Will update proposal with new plan shortly.
